core/src/main/scala/sbt/librarymanagement/SemanticSelector.scala

```diff
@@ -0,0 +1,259 @@
+package sbt.librarymanagement
+
+/**
+ * Semantic version selector API to check if the VersionNumber satisfies
+ * conditions described by semantic version selector.
+ */
+sealed abstract case class SemanticSelector(
```

Comment:
For binary compatibility reasons, let's avoid case classes.
I think we should use Contraband here. I have a [PR](https://github.com/sbt/sbt/pull/4139) on sbt/sbt that adds `JavaVersion` for example:

```graphql
type JavaVersion {
  numbers: [Long]
  vendor: String

  #x def numberStr: String = numbers.mkString(".")
  #xtostring vendor.map(_ + "@").getOrElse("") + numberStr

  #xcompanion def apply(version: String): JavaVersion = sbt.internal.CrossJava.parseJavaVersion(version)
}
```